### PR TITLE
Add gradient accumulation to Llama3 recipe 

### PR DESCRIPTION
**Pull request by savitha-eng**
### Description

Implements gradient accumulation for the Llama3 Native TE recipe, following the pattern from ESM2 PR #1254.

This enables training with larger effective batch sizes without increasing GPU memory usage by accumulating gradients across multiple microbatches before performing an optimizer step.

**Key Changes:**
- **perf_logger.py**: Added `log_micro_step()` method to track metrics across microbatches, updated `log_step()` signature to use accumulated metrics, added configurable `pad_token_id` parameter (defaults to 1)
- **train_ddp.py**: Implemented gradient accumulation loop with `model.no_sync()` for efficiency, added validation for `grad_acc_steps >= 1`
- **train_fsdp2.py**: Implemented gradient accumulation loop (without `model.no_sync()` as FSDP2 handles synchronization internally), added validation
- **defaults.yaml**: Added `grad_acc_steps` parameter (default: 1 for backward compatibility)
- **test_gradient_accumulation.py**: Added golden value test that validates mathematical correctness of gradient accumulation

**Validation:**
Lingua1B DCLM Benchmark trained with Grad_Acc=4, 2 Nodes, MBS=4 -> GBS=256 https://api.wandb.ai/links/clara-discovery/5laqf4gm
Has matching loss curve: 
image

DDP Results: https://api.wandb.ai/links/clara-discovery/6ncxn9n4
- DDP Training Loss curves for single node & 4 node training runs are similar with varying levels of gradient accumulation (grad acc=1, grad acc=2, grad acc=4) for a mbs=4: 
image

FSDP2 Results: https://api.wandb.ai/links/clara-discovery/lcvrsgm8

- FSDP2 Training Loss Curves for single node and 4 node training runs are similar with and without gradient accumulation: 
image

Golden value test confirms that `micro_batch=1, grad_acc=2` produces mathematically identical gradients to `micro_batch=2, grad_acc=1`.

**References:**
Adapts the gradient accumulation implementation from ESM2: #1254

##### Without gradient accumulation (default, backward compatible)
python train_fsdp2.py --config-name L2_lingua_1b

##### With gradient accumulation (reduce memory usage)
python train_fsdp2.py \
    --config-name L2_lingua_1b \
    dataset.micro_batch_size=2 \
    grad_acc_steps=2

##### Effective batch size formula:
effective_batch = micro_batch_size × num_gpus × grad_acc_steps
##### Example: 2 × 16 × 2 = 64 samples per optimizer step**Benefits:**
- Enables larger effective batch sizes on memory-constrained GPUs
- Allows training larger models by reducing micro batch size
- Maintains identical training dynamics to larger microbatches
- Backward compatible: `grad_acc_steps=1` behaves as before

#### Type of changes

- [x] New feature (non-breaking change which adds functionality)

### CI Pipeline Configuration

- [ciflow:all-recipes](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all-recipes) - Run tests for all recipes to validate gradient accumulation doesn't break existing tests

### Pre-submit Checklist

- [x] I have tested these changes locally (single-GPU validation on SLURM)
- [x] I have updated the documentation accordingly (inline comments, test docstrings)
- [x] I have added/updated tests as needed (test_gradient_accumulation.py with golden value test)
- [x] All existing tests pass successfully (pre-commit hooks pass, golden value test passes)

### Testing Notes

**Golden Value Test:**
pytest bionemo-recipes/recipes/llama3_native_te/tests/test_gradient_accumulation.py -vValidates that gradient accumulation produces mathematically equivalent gradients by comparing:
- Loss values (within 1% tolerance)
- Gradient norms (within 1% tolerance)  
- Individual parameter gradients (within 0.1% tolerance)

**Integration Testing:**
Testing with Lingua-1B benchmark on DCLM dataset - loss curves match

### CODE REVIEW

File: bionemo-recipes/recipes/llama3_native_te/tests/test_gradient_accumulation.py
Comment (pstjohn) [inline bionemo-recipes/recipes/llama3_native_te/tests/test_gradient_accumulation.py:1]: hmm, this doesn't really test any of our code; it's just a general demo that grad accumulation _can_ work.

if we can't get golden value tests easily, it would be great to at least make sure we have grad_acc 2 in some of our test_train.py tests

File: bionemo-recipes/recipes/llama3_native_te/train_ddp.py [hunk 1]
```diff
@@ -45,6 +46,12 @@ def main(args: DictConfig) -> float | None:
     Returns:
         float: The loss value for the final batch.
     """
+    # Validate arguments.
+    if not args.grad_acc_steps >= 1:
+        raise ValueError(
+            f"Gradient accumulation steps must be an integer greater than or equal to 1, but got: {args.grad_acc_steps}"
+        )
+
```
Comment (pstjohn) [inline bionemo-recipes/recipes/llama3_native_te/train_ddp.py]: we don't need this; it's extra verbosity and we can just trust our users to be adults

File: bionemo-recipes/recipes/llama3_native_te/train_ddp.py [hunk 2]
```diff
@@ -115,54 +122,66 @@ def main(args: DictConfig) -> float | None:
         start_step = 0
         epoch = 0
 
-    perf_logger = PerfLogger(dist_config, args)
+    # Get padding token from config for logging
+    config_for_pad_token = config if hasattr(config, "pad_token_id") else None
+    pad_token_id = config_for_pad_token.pad_token_id if config_for_pad_token else 1
```
Comment (pstjohn) [inline bionemo-recipes/recipes/llama3_native_te/train_ddp.py]: why not just grab it from the tokenizer if it exists? otherwise default to 1. But that may not be a super safe assumption. can we not use the attention mask instead?

File: bionemo-recipes/recipes/llama3_native_te/perf_logger.py [hunk 3]
```diff
@@ -80,44 +83,60 @@ def __init__(self, dist_config: DistributedConfig, args: DictConfig):
...
+        """Store data on micro step for gradient accumulation metrics.
+
+        Args:
+            batch: The batch of data for the micro step.
+            outputs: The outputs of the micro step.
+        """
+        self.grad_acc_step_count += 1
+        self.num_tokens += batch["input_ids"].numel()
+        # Use padding token to count unpadded tokens (same as ESM2)
+        self.num_unpadded_tokens += batch["input_ids"][batch["input_ids"] != self.pad_token_id].numel()
```
Comment (pstjohn) [inline bionemo-recipes/recipes/llama3_native_te/perf_logger.py]: why cant we use attention mask like we did before?

File: bionemo-recipes/recipes/llama3_native_te/train_fsdp2.py
Comment (pstjohn) [inline bionemo-recipes/recipes/llama3_native_te/train_fsdp2.py:1]: same comments on this file as in train_ddp

File: bionemo-recipes/recipes/llama3_native_te/tests/test_train.py [hunk 4]
```diff
@@ -92,6 +112,27 @@ def test_sanity_convergence_ddp_hf(tmp_path, recipe_path):
     assert final_loss < 2.0, f"Final loss {final_loss} is too high, expected < 2.0"
 
 
+def test_sanity_convergence_ddp_hf_grad_acc(tmp_path, recipe_path):
```
Comment (pstjohn) [inline bionemo-recipes/recipes/llama3_native_te/tests/test_train.py]: could probably go without this one

File: bionemo-recipes/recipes/llama3_native_te/tests/test_train.py [hunk 5]
```diff
@@ -175,6 +260,28 @@ def test_sanity_convergence_fsdp2_hf(tmp_path, recipe_path):
     assert final_loss < 2.0, f"Final loss {final_loss} is too high, expected < 2.0"
 
 
+def test_sanity_convergence_fsdp2_hf_grad_acc(tmp_path, recipe_path):
```
Comment (pstjohn) [inline bionemo-recipes/recipes/llama3_native_te/tests/test_train.py]: and this one

**Review by pstjohn — APPROVED** [review]
LGTM!
